src/main/kotlin/com/terraformation/backend/tracking/db/T0Store.kt

```diff
@@ -540,22 +457,168 @@ class T0Store(
+                  PLANTING_SUBZONE_POPULATIONS.SPECIES_ID.`as`("species_id"),
+              )
+              .from(this)
+              .join(PLANTING_SUBZONE_POPULATIONS)
+              .on(PLANTING_SUBZONE_ID.eq(PLANTING_SUBZONE_POPULATIONS.PLANTING_SUBZONE_ID))
+              .where(PLANTING_SITE_ID.eq(plantingSiteId))
+              .unionAll(
+                  DSL.select(
+                          ID.`as`("plot_id"),
+                          PLANTING_ZONE_T0_TEMP_DENSITIES.SPECIES_ID.`as`("species_id"),
+                      )
+                      .from(this)
+                      .join(plantingSubzones)
+                      .on(PLANTING_SUBZONE_ID.eq(plantingSubzones.ID))
+                      .join(PLANTING_ZONE_T0_TEMP_DENSITIES)
+                      .on(
+                          plantingSubzones.PLANTING_ZONE_ID.eq(
+                              PLANTING_ZONE_T0_TEMP_DENSITIES.PLANTING_ZONE_ID
+                          )
+                      )
+                      .where(PLANTING_SITE_ID.eq(plantingSiteId))
+              )
+              .asTable("plot_species")
+
+      val tempPlotIdField = tempPlotSpecies.field("plot_id", ID.dataType)
+      val tempSpeciesIdField = tempPlotSpecies.field("species_id", SpeciesId::class.java)!!
+
+      val tempPlotData =
+          dslContext
+              .selectDistinct(
+                  plantingSubzones.PLANTING_ZONE_ID,
+                  tempSpeciesIdField,
+                  PLANTING_ZONE_T0_TEMP_DENSITIES.ZONE_DENSITY,
+              )
+              .from(this)
+              .leftJoin(tempPlotSpecies)
+              .on(ID.eq(tempPlotIdField))
+              .leftJoin(PLANTING_ZONE_T0_TEMP_DENSITIES)
+              .on(
+                  plantingSubzones.PLANTING_ZONE_ID.eq(
+                          PLANTING_ZONE_T0_TEMP_DENSITIES.PLANTING_ZONE_ID
+                      )
+                      .and(tempSpeciesIdField.eq(PLANTING_ZONE_T0_TEMP_DENSITIES.SPECIES_ID))
+              )
+              .leftJoin(PLANTING_SUBZONE_POPULATIONS)
+              .on(
+                  PLANTING_SUBZONE_ID.eq(PLANTING_SUBZONE_POPULATIONS.PLANTING_SUBZONE_ID)
+                      .and(tempSpeciesIdField.eq(PLANTING_SUBZONE_POPULATIONS.SPECIES_ID))
+              )
+              .leftJoin(PLANTING_SUBZONES)
+              .on(PLANTING_SUBZONE_POPULATIONS.PLANTING_SUBZONE_ID.eq(PLANTING_SUBZONES.ID))
+              .where(PLANTING_SITE_ID.eq(plantingSiteId))
+              .and(IS_AD_HOC.eq(false))
+              .and(PERMANENT_INDEX.isNull)
+              .fetch { record ->
+                T0SiteTempPlotData(
+                    plantingZoneId = record[plantingSubzones.PLANTING_ZONE_ID.asNonNullable()],
+                    speciesId = record[tempSpeciesIdField],
+                    zoneDensity = record[PLANTING_ZONE_T0_TEMP_DENSITIES.ZONE_DENSITY],
+                )
+              }
+
+      return tempPlotData.all { it.zoneDensity != null }
```

Comment:
Similar to my comment above: the "is there any entry with null density?" logic could be baked into the SQL.